Feat-dynamic-logo

joshuaunity opened the pull request:
## Description

- Modification to Account Model to include ```logo_url```
- Modification to flexmeasures add account command to accept a logo url
- cleared some unused code

## Look & Feel

Here is an account created with the new additions
flexmeasures add account --name JGreenOfficial --logo-url "https://cloud.seita.nl/index.php/apps/files_sharing/publicpreview/G6FdZg9i3xyqRmY?file=/&fileId=586551&x=1920&y=1080&a=true&etag=ffb7f99ea577188d3a427f4eb71e90a9"

![Screenshot](https://cloud.seita.nl/index.php/apps/files_sharing/publicpreview/PXbonpTadXyQ6TG?file=/&fileId=586591&x=1920&y=1080&a=true&etag=d7aaff68993c9854f845562388682ca8)

## How to test

Run the below command using the new additions to the ``` flexmeasures add account ``` command

**_Command Blueprint:_**
`flexmeasures add account --name <company-name> --logo-url "<img_url>"`

**_Full example:_**
`flexmeasures add account --name MrcEnergy --logo-url "https://cloud.seita.nl/index.php/apps/files_sharing/publicpreview/G6FdZg9i3xyqRmY?file=/&fileId=586551&x=1920&y=1080&a=true&etag=ffb7f99ea577188d3a427f4eb71e90a9" `

## Further Improvements
None

## Related Items

This PR Closes the issue: #471 

---

- [✅] I agree to contribute to the project under Apache 2 License. 
- [✅] To the best of my knowledge, the proposed patch is not based on code under GPL or other license that is incompatible with FlexMeasures

Flix6x commented:
Please note that I changed the base branch for this PR, so the revisions don't also include those from the account colors PR. For one, this makes for a better reviewing experience.

nhoening commented:
I asked Joshua if this was intentional and he said yes, he is relying on something from the other PR here.
He also didn't ask for our  reviews yet ...

Flix6x commented:
The PR rebase is not a commit and doesn't affect any functionality. It just affects the diff that is visible on the PR, and the action that happens upon merging, namely, it would now merge to the `dynamic-account-colors` branch rather than `main`.

I consider my comment to be more about teaching repository maintenance than code review. Setting the base branch correctly for a PR that itself depends on another dev branch is not always known by everyone.

joshuaunity commented:
> The PR rebase is not a commit and doesn't affect any functionality. It just affects the diff that is visible on the PR, and the action that happens upon merging, namely, it would now merge to the `dynamic-account-colors` branch rather than `main`.
> 
> I consider my comment to be more about teaching repository maintenance than code review. Setting the base branch correctly for a PR that itself depends on another dev branch is not always known by everyone.


Hi @Flix6x,

Thank you for the feedback. This PR was created to keep a record of the work in progress while I was addressing issues on the dynamic-colors-branch. At the time, it wasn't ready for review, so I left it as-is, intending to revisit it later.

I appreciate the insights on repository maintenance. Setting the base branch correctly for a PR that depends on another development branch is indeed important and something I'll keep in mind moving forward. Thanks for pointing that out!

Flix6x commented on flexmeasures/cli/data_edit.py:
```diff
@@ -208,7 +208,7 @@ def resample_sensor_data(
     skip_integrity_check: bool = False,
 ):
-    """Assign a new event resolution to an existing sensor and resample its data accordingly."""
+    """Assign aaaa new event resolution to an existing sensor and resample its data accordingly."""
```
typo

joshuaunity replied:
oops, will fix that

Flix6x commented on flexmeasures/cli/utils.py:
```diff
@@ -316,3 +316,29 @@ def validate_color_hex(ctx, param, value):
+def validate_url(ctx, param, value):
+    """
+    Validates that a given value is a valid URL format using regex.
+
```
Assuming you didn't write the regex from scratch, please cite its source in the docstring in accordance with the licence set by the publisher. For instance, the StackOverflow licence demands a citation.

joshuaunity replied:
You're right, I didn't write most of these myself. After creating the initial validation, I was experimenting with different regex patterns and got some helpful suggestions from GitHub Copilot. Although I didn't need the majority of them, I'll take those out and keep only the important one.

Flix6x commented on flexmeasures/cli/data_add.py:
```diff
@@ -174,6 +174,11 @@ def new_account_role(name: str, description: str):
 )
+@click.option(
+    "--logo-url",
+    callback=validate_url,
```
The db revision file suggests the max length is 255 characters. Let's add validation for that.

Also, maybe we should demand shorter urls.

joshuaunity replied:
Hmm, ok will work on that

Flix6x commented on flexmeasures/data/migrations/versions/ea84cefba5c0_added_logo_url_field_to_account_model.py:
```diff
@@ -0,0 +1,31 @@
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Please check the commands and remove the comments.

joshuaunity replied:
sure 👍

Flix6x commented on flexmeasures/ui/utils/color_defaults.py:
```diff
@@ -10,7 +10,7 @@ def get_color_settings(account: Account | None) -> dict:
 
     primary_color: str = "#1a3443"
-    secondary_color: str = "f1a122"
+    secondary_color: str = "#f1a122"
```
Looks like a bug potentially worth a separate PR in case this PR still takes a while to get over the fence.

nhoening replied:
This PR will not take long.
I confirmed that it works, we have only a few things to go. No need for a spearate PR I think.

Flix6x commented on flexmeasures/ui/utils/view_utils.py:
```diff
@@ -86,14 +87,19 @@ def render_flexmeasures_template(html_filename: str, **variables):
+    )
+
+    # check if user has logo_url set
+    if account and account.logo_url:
```
Account would be None only when visiting the login page, right? Perhaps in addition to a per-account logo we could have a server-default logo different to the FlexMeasures logo. Just an idea for a potential follow-up ticket.

nhoening replied:
We do have that, it's the config setting `FLEXMEASURES_MENU_LOGO_PATH`

Flix6x replied:
Oh right, of course.

nhoening replied:
But indeed, the user experience of white-labeling will be not 100% perfect, as it only begins after authentication.

Probably okay for now. I guess the real way out would be to offer partners their own subdomain and checking the subdomain used in our code, which then can be linked to the logo of an account with that name ...  Too much work for now.

nhoening commented on flexmeasures/ui/utils/view_utils.py:
```diff
@@ -86,14 +87,19 @@ def render_flexmeasures_template(html_filename: str, **variables):
+    )
+
+    # check if user has logo_url set
+    if account and account.logo_url:
```
Here, we should have similar logic as with the colors: If the consultancy account has set this, (and I haven't), take theirs.

joshuaunity replied:
ok i can implement that

nhoening reviewed (changes requested):
I tested it and it works well (with a logo of a good size, maybe in the future we want to help them choose something of a good size). Or can we add this already to the hekp string in the CLI command?

Anyway, I have one request concerning the consultancy account logo to be relevant.

nhoening commented on flexmeasures/cli/utils.py:
```diff
@@ -328,17 +328,17 @@ def validate_url(ctx, param, value):
-        r"^(https?|ftp)://"
-        r"(([A-Za-z0-9$_.+!*\'(),;?&=-]|%[0-9A-Fa-f]{2})+@)?"  # User and pass
+        r"^(https?|ftp)://"  # Protocol: http, https, or ftp
         r"(([A-Za-z0-9.-]+)\.([A-Za-z]{2,6})|"  # Domain name
```
I believe you removed too much from this regex - why not more than just the domain?

Also there seems to be a regex error when I create a new account, so it seems you have not really tested if this works with a real URL. (I used https://smartdodos.com/wp-content/uploads/2021/12/logo.webp)

Here is my error:

```
± flexmeasures add account --name LogoDummy --logo-url https://smartdodos.com/wp-content/uploads/2021/12/logo.webp
(...)
Traceback (most recent call last):
  File "/home/nicolas/envs/fm-3.10/bin/flexmeasures", line 33, in <module>
    sys.exit(load_entry_point('flexmeasures', 'console_scripts', 'flexmeasures')())
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 1686, in invoke
    sub_ctx = cmd.make_context(cmd_name, args, parent=ctx)
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 943, in make_context
    self.parse_args(ctx, args)
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 1408, in parse_args
    value, args = param.handle_parse_result(ctx, opts, args)
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 2400, in handle_parse_result
    value = self.process_value(ctx, value)
  File "/home/nicolas/envs/fm-3.10/lib/python3.10/site-packages/click/core.py", line 2362, in process_value
    value = self.callback(ctx, self, value)
  File "/home/nicolas/workspace/seita/flexmeasures/flexmeasures/cli/utils.py", line 330, in validate_url
    url_regex = re.compile(
  File "/usr/lib/python3.10/re.py", line 251, in compile
    return _compile(pattern, flags)
  File "/usr/lib/python3.10/re.py", line 303, in _compile
    p = sre_compile.compile(pattern, flags)
  File "/usr/lib/python3.10/sre_compile.py", line 788, in compile
    p = sre_parse.parse(p, flags)
  File "/usr/lib/python3.10/sre_parse.py", line 955, in parse
    p = _parse_sub(source, state, flags & SRE_FLAG_VERBOSE, 0)
  File "/usr/lib/python3.10/sre_parse.py", line 444, in _parse_sub
    itemsappend(_parse(source, state, verbose, nested + 1,
  File "/usr/lib/python3.10/sre_parse.py", line 843, in _parse
    raise source.error("missing ), unterminated subpattern",
re.error: missing ), unterminated subpattern at position 16
```

joshuaunity replied:
Apologies for the delay in responding to this comment.

I left the regex patterns in place, as they handle the main validation (protocol and domain).

I'll investigate the error further to resolve it and push an update

nhoening replied:
Apart from the error, I believe only protocol and domain will miss out the path that all URLs have. Some URLs might have a port as well - I'm unsure why you threw out most of it. Felix only suggested to add a link to where we got it from.

joshuaunity replied:
After further investigation, I discovered that the issue was with the regex validation. The additional check for IP addresses, which I had removed, was actually necessary to ensure the overall validation worked correctly. 

Initially, I believed that validating just the domain and protocol would be sufficient, but I underestimated the complexity of the regex. Moving forward, it might be more reliable to use a dedicated package for this validation, but I believe for now this should suffice.

joshuaunity replied:
I did consider URLs with ports but assumed that users would primarily provide standard image links with typical domains. However, if you think there's a chance our users might need to include URLs with ports, I'm happy to add that validation back in.

I didn't include the other patterns because I wasn't entirely clear on how they worked—especially the one related to fragments. The path and query string validations also seemed more like nice-to-have features rather than essentials.

nhoening replied:
Query string might not happen a lot in our URL case here, but the path is essential (image file name would be part of it, see my example). Maybe a URL could even have a query string like `?file=logo.jpg` so if the regex can offer that, I'd keep it as well.

joshuaunity replied:
I see, I’ve included the pattern for the path as well. I'll also keep the query string validation to cover cases like ?file=logo.jpg.

nhoening reviewed (changes requested):
I believe this is broken now.

Flix6x commented:
Don't forget the changelog entry.

joshuaunity commented:
> Don't forget the changelog entry.

I have updated the changelog with the new update ✅

Flix6x commented on documentation/changelog.rst:
```diff
@@ -12,12 +12,13 @@ New features
 ----------------------
 * Support new single-belief fast track (looking up only one belief) [see `PR #1067 <https://github.com/FlexMeasures/flexmeasures/pull/1067>`_]
-* Add new annotation types: ``"error"`` and ``"warning"`` [see `PR #1131 <https://github.com/FlexMeasures/flexmeasures/pull/1131>`_]
+* Add new annotation types: ``"error"`` and ``"warning"`` [see `PR #1145 <https://github.com/FlexMeasures/flexmeasures/pull/1145>`_]
```
Please fix this mistake.

nhoening commented on flexmeasures/ui/utils/view_utils.py:
```diff
@@ -86,14 +87,22 @@ def render_flexmeasures_template(html_filename: str, **variables):
+    )
+
+    # check if user/consultant has logo_url set
+    if account:
```
Now, when we don't have an account, FLEXMEASURES_MENU_LOGO_PATH will not be applied. I believe applying it (like before) should be the default, and potentially setting the logo from account or consultancy account happens as the next step.

joshuaunity replied:
Sure, I can do that. But could you share a scenario where we might not have an account? The only instance that comes to mind is the login page, since accessing anything else would typically require an account.

nhoening replied:
True, at this moment that is the situation that would not work anymore.

joshuaunity replied:
Ok, will make an update soon

joshuaunity replied:
Hi @nhoening this had been worked on, but I don't know why its still not showing here

nhoening replied:
Maybe not in this conversation view, but I see it in the overall view lines 102 and 103 :)

joshuaunity replied:
Yeah, i also see it when i go to the branch directly, only when I click on the file from here that I cant see it.

nhoening reviewed (commented):
I found a bug which would stop applying FLEXMEASURES_MENU_LOGO_PATH in some situations.

Flix6x commented:
https://github.com/FlexMeasures/flexmeasures/pull/1150 is about to be merged, and contains a db revision on top of the same previous revision that this PR uses. Running `flexmeasures db merge` should create a merge revision file that resolves the diverging db revisions.

joshuaunity commented:
> #1150 is about to be merged, and contains a db revision on top of the same previous revision that this PR uses. Running `flexmeasures db merge` should create a merge revision file that resolves the diverging db revisions.

alright will work on consolidating the migrations 👍

joshuaunity commented:
> #1150 is about to be merged, and contains a db revision on top of the same previous revision that this PR uses. Running `flexmeasures db merge` should create a merge revision file that resolves the diverging db revisions.

Hi @Flix6x the migrations have been resolved

nhoening commented:
Hm. Can you explain how you resolved it @joshuaunity ?

I had expected a merge revision which has two revisions as `down_revision` (that is what I believe `flexmeasures db merge` creates).

Now you pointed your revision to revise `ea84cefba5c0`, which I am not sure what it is, and added an empty revision on top of your own revision (`763dd9b6ca45`) - which is not having an effect I think.

Did you merge main before you did what you did (and include PR #1150)?

joshuaunity commented:
> Hm. Can you explain how you resolved it @joshuaunity ?
> 
> I had expected a merge revision which has two revisions as `down_revision` (that is what I believe `flexmeasures db merge` creates).
> 
> Now you pointed your revision to revise `ea84cefba5c0`, which I am not sure what it is, and added an empty revision on top of your own revision (`763dd9b6ca45`) - which is not having an effect I think.
> 
> Did you merge main before you did what you did (and include PR #1150)?

After running `flexmeasures db merge` I added the two migrations to the down_revision, I ended up with two heads in my migration history, even after running `flexmeasures db upgrade`. I believe this shouldn't happen. To resolve it, I manually reconnected the merge migration(`add_warning`)  to the new migration file and then connected that to my `add_logo` migration file.

This then brought me to have only one head in the history.

nhoening commented:
> > Hm. Can you explain how you resolved it @joshuaunity ?
> > I had expected a merge revision which has two revisions as `down_revision` (that is what I believe `flexmeasures db merge` creates).
> > Now you pointed your revision to revise `ea84cefba5c0`, which I am not sure what it is, and added an empty revision on top of your own revision (`763dd9b6ca45`) - which is not having an effect I think.
> > Did you merge main before you did what you did (and include PR #1150)?
> 
> After running `flexmeasures db merge` I added the two migrations to the down_revision, I ended up with two heads in my migration history, even after running `flexmeasures db upgrade`. I believe this shouldn't happen. To resolve it, I manually reconnected the merge migration(`add_warning`) to the new migration file and then connected that to my `add_logo` migration file.
> 
> This then brought me to have only one head in the history.

Well, something did not go well.
Check flexmeasures/data/migrations/versions/5a9473a817cb_merge_a60cc43aef5e_and_521034349543.py for an example.

We can also create tha manually.

joshuaunity commented:
> > > Hm. Can you explain how you resolved it @joshuaunity ?
> > > I had expected a merge revision which has two revisions as `down_revision` (that is what I believe `flexmeasures db merge` creates).
> > > Now you pointed your revision to revise `ea84cefba5c0`, which I am not sure what it is, and added an empty revision on top of your own revision (`763dd9b6ca45`) - which is not having an effect I think.
> > > Did you merge main before you did what you did (and include PR #1150)?
> > 
> > 
> > After running `flexmeasures db merge` I added the two migrations to the down_revision, I ended up with two heads in my migration history, even after running `flexmeasures db upgrade`. I believe this shouldn't happen. To resolve it, I manually reconnected the merge migration(`add_warning`) to the new migration file and then connected that to my `add_logo` migration file.
> > This then brought me to have only one head in the history.
> 
> Well, something did not go well. Check flexmeasures/data/migrations/versions/5a9473a817cb_merge_a60cc43aef5e_and_521034349543.py for an example.
> 
> We can also create that manually.

Alright, I'll review the sample and see if I can achieve the same result.

joshuaunity commented:
> > > > Hm. Can you explain how you resolved it @joshuaunity ?
> > > > I had expected a merge revision which has two revisions as `down_revision` (that is what I believe `flexmeasures db merge` creates).
> > > > Now you pointed your revision to revise `ea84cefba5c0`, which I am not sure what it is, and added an empty revision on top of your own revision (`763dd9b6ca45`) - which is not having an effect I think.
> > > > Did you merge main before you did what you did (and include PR #1150)?
> > > 
> > > 
> > > After running `flexmeasures db merge` I added the two migrations to the down_revision, I ended up with two heads in my migration history, even after running `flexmeasures db upgrade`. I believe this shouldn't happen. To resolve it, I manually reconnected the merge migration(`add_warning`) to the new migration file and then connected that to my `add_logo` migration file.
> > > This then brought me to have only one head in the history.
> > 
> > 
> > Well, something did not go well. Check flexmeasures/data/migrations/versions/5a9473a817cb_merge_a60cc43aef5e_and_521034349543.py for an example.
> > We can also create that manually.
> 
> Alright, I'll review the sample and see if I can achieve the same result.

@nhoening this has been resolved, I was on a call with @Flix6x and he was able to walk me through it and successfully merged the migrations

nhoening reviewed (approved):
Thanks!
